frontend/packages/dev-console/src/components/catalog/service/CatalogServiceProvider.tsx

```diff
@@ -0,0 +1,90 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { CatalogExtensionHookOptions, CatalogItem } from '@console/plugin-sdk';
+import { ResolvedExtension } from '@console/dynamic-plugin-sdk/src/api/useResolvedExtensions';
+import { keywordCompare } from '../utils/utils';
+import useCatalogExtensions from '../hooks/useCatalogExtensions';
+import CatalogItemsLoader from './CatalogItemsLoader';
+
+export type CatalogService = {
+  type: string;
+  items: CatalogItem[];
+  itemsMap: { [key: string]: CatalogItem[] };
+  loaded: boolean;
+  loadError: any;
+  searchCatalog: (query: string) => CatalogItem[];
+  catalogExtensions: ResolvedExtension<any>;
+};
+
+type CatalogServiceProviderProps = {
+  namespace: string;
+  catalogType?: string;
+  children: (service: CatalogService) => React.ReactNode;
+};
+
+const CatalogServiceProvider: React.FC<CatalogServiceProviderProps> = ({
+  catalogType,
+  children,
+  namespace,
+}) => {
+  const defaultOptions: CatalogExtensionHookOptions = { namespace };
+  const [catalogTypeExtensions, catalogProviderExtensions] = useCatalogExtensions(catalogType);
+
+  const catalogItems = React.useRef<CatalogItem[]>([]);
+  const [catalogItemsMap, setCatalogItemsMap] = React.useState<{ [key: string]: CatalogItem[] }>(
+    {},
+  );
+  const [loadError, setLoadError] = React.useState();
+
+  const loaded =
+    catalogTypeExtensions.length === 0 ||
+    catalogTypeExtensions.every(({ properties: { type } }) => catalogItemsMap[type]);
+
+  React.useEffect(() => {
+    if (loaded) {
+      catalogItems.current = _.sortBy(_.flatten(Object.values(catalogItemsMap)), 'name');
+    }
+  }, [catalogItemsMap, loaded]);
```

Comment:
Setting a ref won't cause a re-render. This should be computed in a memo:
```suggestion
  const catalogItems = React.useMemo(() => loaded ? _.sortBy(_.flatten(Object.values(catalogItemsMap)), 'name') : [], [catalogItemsMap, loaded]);
```

Reply:
Done.